### PR TITLE
Store content in the Redis cache

### PR DESCRIPTION
**Pull request by lubosmj**
ref #496
Required PR: https://github.com/pulp/pulpcore/pull/1751

### CODE REVIEW

**Comment by lubosmj** [comment]
The PR is once again ready for review. HEAD requests are being now cached.

**Comment by lubosmj** [comment]
@gerrod3, would you mind reviewing this PR from the caching perspective?

File: pulp_container/tests/functional/api/test_content_cache.py
Comment (gerrod3) [inline pulp_container/tests/functional/api/test_content_cache.py:183]: Some leftover print statements

File: pulp_container/app/registry.py
Comment (gerrod3) [inline pulp_container/app/registry.py:84]: I think these should error since it should try to await the auth and base_key methods which are non-async.

File: pulp_container/app/cache.py
Comment (ipanova) [inline pulp_container/app/cache.py:27]: I understand that `getall` will get all of the occurrences of `Accept` header, which can be multiple,however, every Accept header can contain a comma separated list of headers.
see more for the reference https://github.com/pulp/pulp_container/blob/main/pulp_container/app/utils.py#L16

Comment (lubosmj) [reply pulp_container/app/cache.py:27]: As far as I know, this had to be made because of how `MultiDict` stores its keys. I focused only on Accept headers that looked like this:
```
Accept: application/vnd.docker.distribution.manifest.list.v2+json,application/vnd.docker.distribution.manifest.v2+json
```

The call converts the header back to its original value from:
```
<CIMultiDictProxy( 'Accept': 'application/vnd.docker.distribution.manifest.list.v2+json',  'Accept': 'application/vnd.docker.distribution.manifest.v2+json')>
```

Are you sure a user can pass multiple Accept headers to a request?

File: pulp_container/app/cache.py
Comment (lubosmj) [reply pulp_container/app/cache.py:47]: Could you please elaborate on the question? The value is accessed, as it should be with respect to the synchronous context.

File: pulp_container/app/cache.py
Comment (lubosmj) [reply pulp_container/app/cache.py:116]: No, the whole for loop can be removed. Thank you for noticing that.

Comment (gerrod3) [reply pulp_container/app/cache.py:116]: Yeah this code that was copied from the content app which was designed for the fact that we don't know where the base_path ends and the plugin's routing begins in the content app. However, pulp_container does know exactly what the base_path is so this loop isn't need and you can just check if the base_path exists in the cache.

File: pulp_container/app/cache.py
Comment (ipanova) [reply pulp_container/app/cache.py:47]: i was asking why line 47 and 32 are different

File: pulp_container/app/cache.py
Comment (ipanova) [reply pulp_container/app/cache.py:27]: based on the http rfc i read time ago - yes, one can pass multiple headers in a request

File: pulp_container/app/cache.py
Comment (lubosmj) [reply pulp_container/app/cache.py:47]: Because we are fetching the headers from different contexts. One within `AsyncContentCache` and another one within `SyncContentCache` (we have different types of request objects).

Comment (lubosmj) [reply pulp_container/app/cache.py:47]: These are two different types of request objects which depends on the context in which they are used. One is of type `aiohttp.request` and the second one is of type `django.http.request`. To access headers, it is necessary to use different attributes provided by that object type.

I am still required to verify whether these two types are correctly handling multiple accept headers that are provided within a single request. Thank you for the comment.

File: pulp_container/tests/functional/api/test_repositories_list.py
Comment (ipanova) [inline pulp_container/tests/functional/api/test_repositories_list.py:70]: with your refactor we're losing this assertion

File: pulp_container/tests/functional/utils.py
Comment (ipanova) [inline pulp_container/tests/functional/utils.py:270]: make it more generic s/latest_image_url/registry_endpoint_url/ or similar

**Review by ipanova — APPROVED** [review]
I don't want to hold back this PR. please address last 2 comments of mine and merge! thank you for working on this :seal:

**Comment by lubosmj** [comment]
@ipanova, this PR can be merged after merging https://github.com/pulp/pulpcore/pull/1751. Do we need to bump the pulpcore requirement to publish this change? I am not sure if it will cause some problems in the future when we do not update the version (since we wanted to move the pulpcore requirement by every deprecation change only (3.20 now), is that right?).

**Pull request merged**
